### PR TITLE
[BACKPORT] Ethernet : changes were made to support ethernet

### PR DESCRIPTION
**Pull request by garfieldG**
-unicast
-broadcast
-multicast
-no SDIO lines

### CODE REVIEW

**Comment by dagar** [comment]
Can you look at the build failures?

**Comment by garfieldG** [comment]
@dagar No build failures now:)

**Comment by dagar** [comment]
Just to be clear, all of this code actually exists in upstream NuttX (master)?

**Comment by garfieldG** [comment]
@dagar No, not all of it, most of it

**Comment by dagar** [comment]
> @dagar No, not all of it, most of it

We need to be clear on that, otherwise the non upstream changes might be lost when we update NuttX.

Does it correspond with a small set of commits that could actually be cherry-picked back?

**Comment by garfieldG** [comment]
No, actually there is just [one big commit](https://github.com/PX4-NuttX/nuttx/pull/27/commits/ae2218f973cdf422927f127dda717789fce2602f) with all changes.
There is the [second commit](https://github.com/PX4-NuttX/nuttx/pull/27/commits/520d9d3046999588bb5a61f4ad4ede1c704d65b8) that can be cherry-picked later on, but it's a very small change and just an addition to the big commit.
What do you suggest? Is there something I can do now to change it?

**Comment by dagar** [comment]
As discussed please split into 1 commit then is a clean backport from upstream NuttX that we can merge now. Then we can look at your other changes separately and figure out where they belong.

**Comment by garfieldG** [comment]
@dagar Just now saw that in Firmware you requested to separate to two different PRs.
In this PR I already separated to two different commits before seeing it (remembered we discussed about two different commits, not two different PRs).
Do you want me to split this one also to two different PRs or is it okay to leave it and keep it as two separated commits and one PR?

**Comment by dagar** [comment]
It depends on the actual contents. The portion that's a pure backport from upstream NuttX should be safe to merge immediately. The rest will need a bit more scrutiny and we should consider getting it upstream as well to reduce future maintenance burden. Does that make sense?

File: arch/arm/src/stm32/stm32_eth.c [hunk 1]
```diff
@@ -3430,11 +3430,14 @@ static inline void stm32_ethgpioconfig(FAR struct stm32_ethmac_s *priv)
 #if defined(CONFIG_STM32_MII) || defined(CONFIG_STM32_RMII)
 
   /* MDC and MDIO are common to both modes */
```
Comment (davids5) [inline arch/arm/src/stm32/stm32_eth.c]: @garfieldG - Since you are closer to the problem, I need your help to understand this change.

What was the problem you are trying to solve? What it different in your HW and why are this changes needed?

Comment (garfieldG) [reply arch/arm/src/stm32/stm32_eth.c]: My board has an ethernet switch, not only a PHY device. 
Most Phy devices use MDIO lines and I guess that might be the reason why it wasn't referred to until now, but my board communicates with the switch through SPI.

**Comment by garfieldG** [comment]
Makes total sense.
But my question still stands, I don't really know if you can merge only part of the commits in pull request (My work system is a bit different, so never had to deal with merging a pull request) so I don't know if I should separate the commits to different pull requests.

File: arch/arm/src/stm32/stm32_eth.c (hunk 1 above)
Comment (davids5) [reply arch/arm/src/stm32/stm32_eth.c]: @garfieldG - I am starting. to get the picture. The custom board you have has a PHY that is supported in nuttx but not wired the same. 

Does it uses RMMI?  Where is the SPI code?

Comment (garfieldG) [reply arch/arm/src/stm32/stm32_eth.c]: Yes, Ethernet PHY (usually) has 2 communication links to the MAC (the CPU hardware). Usually the 100Mbit/sec data is over MII protocol (either RMII, MII or GMII). but the control (on/off, 100mbit/10mbit, HD/FD and so on) is done by other lines, the default is MDIO standard. It seems like Nuttx handles this standard (you can see it in stm32_ethernet code). 
When the PHY is a switch MDIO is less relevant, so you need other communication. In our case its a complicated protocol over SPI lines.
The SPI code is not relevant, because the two protocols are independent. We actually don't do much over the SPI, just monitoring the link (up/down and so on)

Comment (garfieldG) [reply arch/arm/src/stm32/stm32_eth.c]: @davids5

**Comment by garfieldG** [comment]
@dagar

File: arch/arm/src/stm32/stm32_eth.c (hunk 1 above)
Comment (davids5) [reply arch/arm/src/stm32/stm32_eth.c]: @garfieldG - I am not familiar enough with what you are trying to accomplish.   What I was trying and failing at understanding so far, is: Is the change you need to make to a driver in NuttX' stmf7 code base a change specific to just your HW are is it something that would benefit upstream Nuttx? For example, a bug fix or a set of CONFIG_xxx that are wrong in upstream. 

We can have a call to discuss this in real time if need be. Skype:david_s5

**Comment by davids5** [comment]
@garfieldG a PR brings in all the commits so we would need to split it. This is possible with branching your branch and using "git rebase -i" to drop the commits. You could also just cherry pick them to 2 new branches.

File: arch/arm/src/stm32/stm32_eth.c (hunk 1 above)
Comment (garfieldG) [reply arch/arm/src/stm32/stm32_eth.c]: @davids5 
As far as I can see, for now, it is specific to our HW (I I didn't see any reference to a board without MDIO lines), I think it's best we'll leave it for now and in the future if there will be any need for this change I'll be here:)

**Comment by garfieldG** [comment]
@davids5 I'll change it so that only the backport commit will be here and I'll delete the second commit for now

**Comment by davids5** [comment]
@garfieldG Perfect!

**Comment by garfieldG** [comment]
@davids5 Hi, is there something new regarding the Ethernet?

**Comment by davids5** [comment]
@garfieldG @dagar was going to run a regression test on HW he has. Once he OK's it, it will be merged.

**Review by dagar — APPROVED** [review]
